[inductor] Inline ComputedBuffer computation when there are no reads

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102277
* __->__ #102000

When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

peterbell10 committed:
[inductor] Inline ComputedBuffer computation when there are no reads

When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

[ghstack-poisoned]

peterbell10 committed:
Don't inline mutated buffers on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Mark buffer mutations consistently on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -1989,12 +1997,16 @@ def unwrap_views(target):
-        return unwrap_views(self.target).layout
+        result = unwrap_views(self.target)
+        assert isinstance(result, Buffer), "MutationLayout must refer to a buffer"
+        return result
```
I was seeing accuracy failures and narrowed it down to certain fallback kernels like scatter fallback not calling `realize_users_of`, so this strictly enforces that creating a `MutationLayout` will mark the buffer as mutated.

lezcano reviewed (approved):
Great stuff. I also like a lot the split between reindexing and making the actual index to allow you to work with the indices on their own.

peterbell10 committed:
Update on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Lint on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Rebase on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

kit1980 commented:
@pytorchbot revert -m "Broke inductor tests https://github.com/pytorch/pytorch/actions/runs/5095190248/jobs/9160028124" -c landrace

kit1980 commented:
See comments to the next PR in the stack https://github.com/pytorch/pytorch/pull/102277 for more details on the revert.

peterbell10 committed:
Update on "[inductor] Inline ComputedBuffer computation when there are no reads"


When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -5185,7 +5238,7 @@ def run(x):
             expected_kernel = 4
         else:
-            expected_kernel = 5
+            expected_kernel = 6
```
so this is an example of the usual candidate to be fused with `aggressive_fusion = True`, but that we're not fusing ATM. Fair enough.

ezyang commented:
I'm guessing this is responsible for a 2% HF overall improvement. Example

<img width="457" alt="image" src="https://github.com/pytorch/pytorch/assets/13564/1dc5f591-61be-4217-a648-5952b4d20b88">

Other candidate PRs:

```
(base) ezyang-mbp:pytorch-tmp ezyang$ git log --oneline af1d437654aa7fd5aa4a5d30ae2f2e9af34c7765..999bae0f54108ffc5b7cf2524a02a83901554b16 torch/_dynamo torch/_inductor .ci
2cc6ae1926d (origin/gh/willfengg/1/base) squash xblock for persistent inner reduction (#102444)
fc0fed36d93 [inductor] fix issue with ops.lookup_seed (#102485)
c6d9a0b9dd4 [inductor] Handle floordiv and remainder in IndexPropagation (#102277)
e4e151d6693 [inductor] Inline ComputedBuffer computation when there are no reads (#102000)
b1bc8aecf5b [inductor] erfinv: CPU/CUDA lowering (#101863)
```

ezyang commented:
Hmm, actually, also decent chance this is due to https://github.com/pytorch/pytorch/pull/102444
